[PyTorch] Add native fast path for transformer encoder inference

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75809

The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

swolchok committed:
[PyTorch] Add native fast path for transformer encoder inference

The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok commented on torch/nn/modules/activation.py:
```diff
@@ -896,6 +896,25 @@ class MultiheadAttention(Module):
+    - add_zero_attn is False
+    - batch_first is True and the input is batched
+    - kdim and vdim are equal to embed_dim
+    - at most one of key_padding_mask and attn_mask is used
+    - neither key_padding_mask nor attn_mask is passed (this restriction will be loosened)
```
whoops, these two items are contradictory, will fix

swolchok commented on torch/nn/modules/activation.py:
```diff
@@ -896,6 +896,25 @@ class MultiheadAttention(Module):
+    - at most one of key_padding_mask and attn_mask is used
+    - neither key_padding_mask nor attn_mask is passed (this restriction will be loosened)
+
+    If the optimized implementation is in use, a NestedTensor can be
+    passed to more represent padding more efficiently than using a
```
needs to spell out that query/key/value are the things that can be NestedTensor

swolchok commented on torch/nn/modules/transformer.py:
```diff
@@ -267,6 +267,19 @@ class TransformerEncoderLayer(Module):
 
+    forward() will use a special optimized implementation if all of the following
+    conditions are met:
+    - All conditions for the fast path of torch.nn.MultiheadAttention
```
this is sloppy writing on my part that doesn't correspond to how it's actually enforced. will fix.

swolchok commented on torch/nn/modules/transformer.py:
```diff
@@ -267,6 +267,19 @@ class TransformerEncoderLayer(Module):
+    conditions are met:
+    - All conditions for the fast path of torch.nn.MultiheadAttention
+    - norm_first is False (this restriction may be loosened in the future)
+    - activation is one of "relu", "gelu", torch.functional.relu, or torch.functional.gelu
```
colon after "of", I think

swolchok commented on torch/nn/modules/transformer.py:
```diff
@@ -267,6 +267,19 @@ class TransformerEncoderLayer(Module):
+    - activation is one of "relu", "gelu", torch.functional.relu, or torch.functional.gelu
+    - neither src_mask nor src_key_padding_mask is passed (this restriction will be loosened)
+
+    If the optimized implementation is in use, a NestedTensor can be
+    passed to represent padding more efficiently than using a padding
```
should specify that `src` is the thing that can be a NestedTensor

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok commented on test/test_native_mha.py:
```diff
@@ -0,0 +1,239 @@
+            q,
+            k,
+            v,
+            need_weights=True,
+            average_attn_weights=False,
```
need to test all 4 combinations of these flags

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

cpuhrsch commented on torch/nn/modules/activation.py:
```diff
@@ -1037,6 +1058,40 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
+                    self.out_proj.bias,
+                    key_padding_mask if key_padding_mask is not None else attn_mask,
+                    need_weights,
+                    average_attn_weights)
```
Throw on the slow path if a NestedTensor is passed, because it doesn't support it.

cpuhrsch replied:
Bonus points for printing what would need to change for the fast path to hit. Then the user can construct a NestedTensor and learn what was passed vs. expected to hit the fast path / get NestedTensor support (since in this case they're the same).

swolchok replied:
> Bonus points for printing what would need to change for the fast path to hit

I've given it some thought, but I can't see how to do that without either duplicating the list of conditions or adding more overhead.

swolchok replied:
figured it out

albanD reviewed (changes requested):
Thanks for sending this!
I feel like this is adding a lot of things though and would be simpler to review if it was splitted more. This will make the reviewers work easier and also easier for you to keep track of the changes that need to happen while keeping the number of comments on the PR manageable.

In particular, some comments from a very cursory look:
- there seems to be API added both via native_function and torch.ops. They seems independent? Also we should justify why we're doing this one way or the other?
- Testing should be done via OpInfo for the most part, I am surprised to see 100s of lines of test with hardcoded values and no OpInfo?
- The doc refers to NestedTensor but that should point to something more specific so that users can use that information
- torch.ops is known to be ludicrously slow (https://github.com/pytorch/pytorch/issues/74943) so most likely should not be used for anything perf sensitive?

swolchok commented:
> would be simpler to review if it was splitted more

1) just the files for MHA are about 1200 lines, so more splitting is not going to get you below that number
2) There are changes in this PR, but this code was already reviewed internally; if you use Phabricator, you can see which lines haven't changed and focus on "is this acceptable in PyTorch core" vs "is this good software"

> there seems to be API added both via native_function and torch.ops. They seems independent? Also we should justify why we're doing this one way or the other?

The torch.ops stuff is hit only in tests. I'm happy to move it to native_functions.yaml if you don't mind further clutter there.

> Testing should be done via OpInfo for the most part, I am surprised to see 100s of lines of test with hardcoded values and no OpInfo?

This is an extension to an existing feature, so I have mostly extended the existing tests, which evidently did not use OpInfo. Converting existing non-OpInfo tests to OpInfo should be out of scope, but I can look into OpInfo for the new tests in test_native_mha.py; can you point me to the doc on how to use it?

> The doc refers to NestedTensor but that should point to something more specific so that users can use that information

I don't know how to do that. Can you be more specific? Are you asking me to add a hyperlink in the docstring? How does that work?

> torch.ops is known to be ludicrously slow (https://github.com/pytorch/pytorch/issues/74943) so most likely should not be used for anything perf sensitive?

used only in tests.

swolchok commented:
> I can look into OpInfo for the new tests

I'm not adding new operators here, just an is_inference_mode_enabled fast path, so I don't see how OpInfo applies.

ezyang commented on aten/src/ATen/native/transformers/attention.h:
```diff
@@ -0,0 +1,16 @@
+    const int64_t num_head);
+void masked_softmax_dropout_cuda(
+    const Tensor& attn_scores,
+    const c10::optional<Tensor>& attn_mask);
```
I think you shouldn't need to explicitly define the header declarations here; you can get them from `NativeFunctions.h` or `ATen/ops/_transform_bias_rescale_qkv.h` (need to case on `AT_PER_OPERATOR_HEADERS`)

swolchok replied:
will try removing

swolchok replied:
looks like we can just delete the header and replace it with nothing.

ezyang commented on aten/src/ATen/native/transformers/cuda/library.cpp:
```diff
@@ -0,0 +1,26 @@
+TORCH_LIBRARY_IMPL(nativetransformers, CUDA, m) {
+  m.impl(
+      "_transform_bias_rescale_qkv_op",
+      TORCH_FN(transform_bias_rescale_qkv_op_cuda));
```
huh, is there a reason these aren't in the aten namespace (and then you could have used `native_functions.yaml` for them)

swolchok replied:
deleted!

ezyang commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -105,6 +105,7 @@ file(GLOB native_quantized_cpp
             "native/quantized/*.cpp"
             "native/quantized/cpu/*.cpp")
 file(GLOB native_nested_cpp "native/nested/*.cpp")
+file(GLOB native_transformer_cpp "native/transformers/*.cpp")
```
I guess I personally wouldn't have bothered making a new folder cuz build system shenanigans, but I guess if you got it to work :)

swolchok replied:
it would be nice if our build system didn't care so much about directory structure. this was much harder to get working than it would have been with buck.

ezyang commented on aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:
```diff
@@ -35,6 +35,12 @@ Tensor nested_from_padded_cuda(
+    if (padded.dtype() != kFloat && padded.dtype() != kHalf) {
+      TORCH_WARN_ONCE(
+          "nested_from_padded CUDA kernels only support fp32/fp16; falling "
+          "back to slower generic kernel");
```
another one for https://github.com/pytorch/pytorch/issues/55655

ezyang commented on aten/src/ATen/native/transformers/transformer.cpp:
```diff
@@ -0,0 +1,136 @@
+    const Tensor& b2,
+    bool use_gelu,
+    bool add_norm) {
+  TORCH_CHECK(add_norm == false, "TODO add_norm to be supported in FFN");
```
it's customary to ask people to report a bug if they want support hee hee ;)

ezyang replied:
I guess this shouldn't actually get hit by the test for eligibility in the normal path

ezyang commented on torch/nn/modules/transformer.py:
```diff
@@ -267,6 +267,24 @@ class TransformerEncoderLayer(Module):
 
+    forward() will use a special optimized implementation if all of the following
+    conditions are met:
+    - inference mode is enabled (using the torch.inference_mode() context manager)
```
The code fragments here will format better if they are surrounded by double backticks

ezyang replied:
I think I mentioned this when we were chatting, but to record it here: I think it's a little surprising that inference mode has to be enabled to hit the optimized implementation; just testing that the inputs don't require grad should be enough.

swolchok replied:
> just testing that the inputs don't require grad should be enough

don't we want people to use torch.inference_mode, not torch.no_grad? this seems like a good way to push them toward the right behavior.

ezyang commented on torch/nn/modules/transformer.py:
```diff
@@ -267,6 +267,24 @@ class TransformerEncoderLayer(Module):
+    forward() will use a special optimized implementation if all of the following
+    conditions are met:
+    - inference mode is enabled (using the torch.inference_mode() context manager)
+    - training is disabled (using .eval())
```
Since this operation is supposed to be semantically equivalent when no gradients are run, I don't think it's necessary to test for eval(). The main reason for eval() is that we have some modules (like dropout) that literally do different things depending on if they're in training or eval mode. This does the same thing (but doesn't support backprop), so the reasoning for this doesn't apply.

swolchok replied:
NestedTensor cannot be used in training because it doesn't support grad, and we don't support dropout in the native implementation, so the fast path definitely will not work in training; does this meet the bar for a .eval() check?

ezyang replied:
OK

ezyang commented on torch/nn/modules/transformer.py:
```diff
@@ -313,16 +331,22 @@ def __init__(self, d_model: int, nhead: int, dim_feedforward: int = 2048, dropou
+        elif activation is F.gelu:
+            self.activation_relu_or_gelu = 2
+        else:
+            self.activation_relu_or_gelu = 0
```
prolly worth a little comment saying that this is to appease TorchScript

swolchok replied:
done!

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok commented on torch/nn/modules/activation.py:
```diff
@@ -1037,6 +1058,43 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
+            ((key_padding_mask is None and attn_mask is None)
+             if query.is_nested
+             else (key_padding_mask is None or attn_mask is None)) and
+            query is key and key is value):
```
why is there not an automated tool fixing this flake8 stuff for me

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

cpuhrsch reviewed (commented):
The NestedTensor bits seem fine to me, but I can't speak to the general test changes, doc string updates and overall code organization.

swolchok commented:
> So we might as well make them different PR.

I would agree with you if this was new code, but this has already been developed in a bunch of separate diffs internally and is logically an `hg mv` into PyTorch core. The kernels do not require detailed re-review from scratch.

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok commented:
Some things making this PR difficult to split:
- 900+ lines just for the MHA kernels, so it will never be small
- test_nn.py is huge; splitting out the (comparatively small and, per Phabricator, mostly unchanged) TransformerEncoderLayer fast path would entail a semi-manual edit in `hg split` that would need to be performed exactly the same way twice thanks to dirsync
- Replicating changes to diff structure across the 3+ different build systems involved is error-prone

In short, the cost in my time is high and the perceived benefit given that the code is not new and Phabricator will show what has been changed is not high.

swolchok committed:
fix CMakeLists, build_amd, and hipify_python on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok commented:
`torch.is_inference_mode_enabled` seems to not work in TorchScript. Not sure why or what to do about it.

ezyang commented:
> torch.is_inference_mode_enabled seems to not work in TorchScript. Not sure why or what to do about it.

I guess you could add support for it in TS... personally I'd probably see if the requires_grad tests were valid TS and do that instead

cpuhrsch commented on test/test_native_mha.py:
```diff
@@ -0,0 +1,288 @@
+        )
+
+
+instantiate_device_type_tests(TestMHADeviceType, globals())
```
Are we sure this is running in CI without adding "run_tests" at the end of the file? For example https://github.com/pytorch/pytorch/blob/master/test/test_torch.py#L8387-L8388

albanD replied:
It does actually. Just that the github view cuts the file before the end.

swolchok replied:
fixed

cpuhrsch commented on test/test_native_mha.py:
```diff
@@ -0,0 +1,288 @@
+    @skipMeta
+    @torch.inference_mode()
+    def test_native_multihead_self_attention(self, device, dtype):
+        for (use_padding, pad_all) in ((False, False), (True, False), (True, True)):
```
You could consider using parametrize here. For example see https://github.com/pytorch/pytorch/blob/master/test/test_sparse_csr.py#L649-L650

swolchok replied:
TIL! this is intentionally not "all combinations" though, because `(False, True)` is the same as `(False, False)`.

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok commented:
Talked to @ezyang offline; decided to check requires_grad and torch.is_grad_enabled rather than torch.is_inference_mode_enabled.

Remaining problems seem to be:
- docs build
- CUDA files can't see NestedTensor_get_max_size and this is persisting even after I put `TORCH_API` on it for some reason

swolchok committed:
rename attention.cu to attention.cpp in hopes of fixing linker errors. hopefully fix docs build. on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok committed:
rename attention.cpp back to attention.cu. hopefully fix docs build again -- apparently blank lines are not ok on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/transformers/attention.cpp:
```diff
@@ -0,0 +1,482 @@
+    attn_mask = at::expand_inplace(attn_scores, *attn_mask)->contiguous();
+  }
+  if (attn_mask) {
+    attn_scores = _masked_softmax(attn_scores, *attn_mask);
```
This doesn't look good, writing over mutable Tensor& ref instead of copying into it

ezyang replied:
Oh maybe it's fine though, `attn_scores` gets passed in as an argument. Better not to also write over it though

ezyang commented on aten/src/ATen/native/transformers/attention.cpp:
```diff
@@ -0,0 +1,482 @@
+  }
+#endif
+
+#ifdef DEBUG_PRINT_EACH_STEP
```
there's no way to set this preproc from OSS, is that right?

ezyang commented on test/test_native_mha.py:
```diff
@@ -0,0 +1,288 @@
+            (24, 4, 4, 2),
+            (48, 4, 16, 8),
+        ]
+        for (embed_dim, num_heads, bs, sl) in tests:
```
nit: you can modestly improve debuggability when a test fails by using subTest to report which parameters were chosen for this test

ezyang reviewed (approved):
did not review the kernels, scaffolding seems fine. Tests are not completely idiomatic (for example, device parametrization should be used for testing on different devices but it's alright).

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok committed:
add subTest, try yet again on docs build on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok committed:
fix outdent in docstring on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok commented:
will send a follow-up PR to fix formatting of bulleted list in fast path docstrings.

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok commented on aten/src/ATen/native/transformers/transformer.cpp:
```diff
@@ -0,0 +1,135 @@
+  return result.view({mat1.sizes()[0], mat1.sizes()[1], -1});
+}
+
+Tensor ffn_cpu(
```
this should just be called `ffn`, whoops

swolchok committed:
add run_tests call to test_native_mha on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok commented:
linux-bionic-py3.7-clang9 test failure is puzzling (gating for the fast path somehow failed only on one platform?), and Dr. CI didn't call it out in the comment, confused about why.

swolchok committed:
add explanation of why the fast path wasn't hit. should help debug unexpected failure. also raise tolerances for test for windows on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

albanD commented on test/test_nn.py:
```diff
@@ -5823,6 +5822,100 @@ def test_multihead_attn_invalid_shape(self):
+        with torch.inference_mode():
+            self._test_multihead_attn_invalid_shape_impl(mha.eval())
+
+    @torch.inference_mode()
```
Should this be `no_grad()` ?

albanD commented on test/test_nn.py:
```diff
@@ -7718,14 +7811,16 @@ def get_a_test_layer(use_cuda, activation, batch_first=False):
 
             model = nn.TransformerEncoder(encoder_layer, 1).to(device)
+            if not training:
+                model = model.eval()
```
you can do `mode.train(trainigng)`.
Also no need to get the result, this is inplace.

albanD commented on test/test_nn.py:
```diff
@@ -7844,7 +7947,15 @@ def perm_fn(x):
+                if training:
+                    cm = contextlib.nullcontext()
+                else:
+                    cm = torch.inference_mode()
```
no_grad

note that inference mode sets no grad internally. So no_grad is stricter

albanD commented on test/test_nn.py:
```diff
@@ -14489,11 +14603,27 @@ def test_Bilinear_empty(self, device):
+                encoder_layer = nn.TransformerEncoderLayer(d_model=512, nhead=8, batch_first=batch_first).to(device)
+                if not training:
+                    encoder_layer = encoder_layer.eval()
+                    with torch.inference_mode():
```
no grad

albanD commented on test/test_nn.py:
```diff
@@ -17924,6 +18054,32 @@ def test_multihead_attention_dtype(self, device, dtype):
+            model = nn.MultiheadAttention(embed_dim, num_heads, batch_first=True).cuda().to(dtype)
+            if not training:
+                model = model.eval()
+                cm = torch.inference_mode()
```
no grad

albanD commented on test/test_nn.py:
```diff
@@ -19902,13 +20058,17 @@ def test_transformerencoderlayer(self, device, dtype):
                                                batch_first=batch_first, device=device, dtype=dtype)
 
+            if not training:
+                assert dropout == 0
```
setting eval mode will disable dropout

albanD commented on test/test_nn.py:
```diff
@@ -19902,13 +20058,17 @@ def test_transformerencoderlayer(self, device, dtype):
 
+            if not training:
+                assert dropout == 0
+                model = model.eval()
```
`model.train(training)`

albanD commented on test/test_nn.py:
```diff
@@ -20001,6 +20163,64 @@ def perm_fn(x):
 
+            # NestedTensor is only supported for the fast path
+            # currently, which won't be used if training.
+            if batch_first and not training and ('cuda' in str(device) or 'cpu' in str(device)):
```
is_cuda, is_cpu

albanD commented on test/test_nn.py:
```diff
@@ -20001,6 +20163,64 @@ def perm_fn(x):
+
+                nt = torch.nested_tensor([encoder_input[0][:-1], encoder_input[1]], device=device)
+                result = model(nt)
+                ref_output = torch.tensor(
```
Couldn't we call the slow path to get these reference instead of hard-coding them with "random" accuracy?

albanD commented on test/test_nn.py:
```diff
@@ -20001,6 +20163,64 @@ def perm_fn(x):
+                )
+                self.assertEqual(tuple(result.shape), tuple(ref_output.shape))
+                if 'cuda' in device:
+                    if dtype == torch.float:
```
Why is this needed?

albanD commented on test/test_nn.py:
```diff
@@ -20001,6 +20163,64 @@ def perm_fn(x):
+                if training:
+                    cm = contextlib.nullcontext()
+                else:
+                    cm = torch.inference_mode()
```
no grad

albanD commented on test/test_nn.py:
```diff
@@ -20017,12 +20237,15 @@ def test_transformerencoderlayer_gelu(self, device, dtype):
 
             model = nn.TransformerEncoderLayer(d_model, nhead, dim_feedforward, dropout,
                                                activation, batch_first=batch_first, device=device, dtype=dtype)
+            if not training:
```
same as above

albanD commented on test/test_native_mha.py:
```diff
@@ -0,0 +1,301 @@
+        ]
+        for (embed_dim, num_heads, bs, sl) in tests:
+            with self.subTest(embed_dim=embed_dim, num_heads=num_heads, bs=bs, sl=sl):
+                torch.manual_seed(9343)
```
The test suite is already setting the seed at the beginning of the test. Why do you need this one?

albanD commented on test/test_native_mha.py:
```diff
@@ -0,0 +1,301 @@
+        for (embed_dim, num_heads, bs, sl) in tests:
+            with self.subTest(embed_dim=embed_dim, num_heads=num_heads, bs=bs, sl=sl):
+                torch.manual_seed(9343)
+                dense_x = x = (
```
Why have both dense_x and x here?
Also since you inplace modify x, you also inplace modify dense_x. Is that expected?

swolchok replied:
x is potentially rebound later (line 41), whereas dense_x is not.

albanD commented on test/test_native_mha.py:
```diff
@@ -0,0 +1,301 @@
+    @dtypesIfCUDA(torch.float, torch.half)
+    @dtypes(torch.float)
+    @skipMeta
+    @torch.inference_mode()
```
no grad

swolchok replied:
I did inference_mode on purpose. Why would we keep leaving examples of no_grad laying around when inference_mode is the better alternative people should default to?

albanD replied:
They are not the same though. Sometimes you want one sometimes the other.
And because inference_mode => no_grad it is better to test no_grad.

swolchok replied:
thanks, this exposed autograd complaining about returning a tensor with use_count() > 1 in test_TransformerEncoderLayer_empty despite being run with no_grad. Not worth fighting autograd over; just cloning the input in that case instead of returning it.

albanD commented on test/test_native_mha.py:
```diff
@@ -0,0 +1,301 @@
+    @dtypesIfCUDA(torch.float, torch.half)
+    @dtypes(torch.float)
+    @skipMeta
+    @torch.inference_mode()
```
no grad

albanD commented on test/test_native_mha.py:
```diff
@@ -0,0 +1,301 @@
+    @dtypesIfCUDA(torch.float, torch.half)
+    @dtypes(torch.float)
+    @skipMeta
+    @torch.inference_mode()
```
no grad

albanD commented on test/test_native_mha.py:
```diff
@@ -0,0 +1,301 @@
+    @dtypes(torch.float)
+    @skipMeta
+    @torch.inference_mode()
+    def test_native_multihead_encoder_decoder_attention(self, device, dtype):
```
nit parametrize over the mode?

swolchok replied:
the non-self modes aren't supported as well, so they aren't tested as thoroughly on purpose. (They are also currently unreachable.) This would be a good idea when it's time to test them all equally, though.

albanD reviewed (commented):
Only looked at the python code.

swolchok commented on torch/nn/modules/activation.py:
```diff
@@ -893,6 +893,27 @@ class MultiheadAttention(Module):
 
+    ``forward()`` will use a special optimized implementation if all of the following
+    conditions are met:
+
```
I definitely spent quite a few round trips on these docstrings. The renderer is extremely picky and difficult to work with, especially from fbcode where I can't build the docs myself to iterate quicker.

swolchok commented on torch/nn/modules/activation.py:
```diff
@@ -893,6 +893,27 @@ class MultiheadAttention(Module):
+    ``query``/``key``/``value`` to more represent padding more efficiently than using a
+    padding mask. In this case, a `NestedTensor <https://pytorch.org/docs/stable/nested.html>`_
+    will be returned, and an additional speedup proportional to the fraction of the input
+    that is padding can be expected.
```
no, I meant padding. speedup = amount of padding / total input size

swolchok commented on torch/nn/modules/activation.py:
```diff
@@ -1034,6 +1056,66 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
+            # We have to use list comprehensions below because TorchScript does not support
+            # generator expressions.
+            if torch.overrides.has_torch_function(tensor_args):
+                why_not_fast_path = "some Tensor argument has_torch_function"
```
why would it be supported? I cannot guarantee at all that torch_functions will be called.

albanD replied:
The codegen that binds your code to python is adding the torch_function handler for you :)

swolchok replied:
I am thinking of torch functions for *other* operators. torch_function is essentially COME FROM. Somebody could use it to get a callback inside MHA, and the fast path will not call those callbacks, which is a behavior change, so we can't use the fast path.

swolchok replied:
Put another way, if somebody was making MHA work by `__torch_function__`ing the pieces it calls into previously, forcing them to execute the fast path will break their code.

swolchok replied:
you can't torch_function all of MHA. it makes no sense to torch_function just the fast path. If you enter the fast path using some kind of weird tensor-like that has torch_functions, it is going to crash. Therefore, we cannot enter the fast path in that case at all; if we tried to, it would break code that previously worked.

swolchok commented on torch/nn/modules/activation.py:
```diff
@@ -893,6 +893,27 @@ class MultiheadAttention(Module):
+    - if a `NestedTensor <https://pytorch.org/docs/stable/nested.html>`_ is passed, neither ``key_padding_mask`` nor ``attn_mask`` is passed
+
+    If the optimized implementation is in use, a
+    `NestedTensor <https://pytorch.org/docs/stable/nested.html>`_ can be passed for
```
it will exist in stable when this hits stable, though. Do we really want to have to write another PR when that happens?

swolchok replied:
next release = 1.12, right?

gchanan replied:
This shouldn't be release specific -- if you put stable it will always point to stable even if that's out of date.  If you put a version number we'll need to remember to update it, which won't happen.

swolchok commented on torch/nn/modules/activation.py:
```diff
@@ -1034,6 +1056,66 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
+            # generator expressions.
+            if torch.overrides.has_torch_function(tensor_args):
+                why_not_fast_path = "some Tensor argument has_torch_function"
+            elif not all([(x.is_cuda or 'cpu' in str(x.device)) for x in tensor_args]):
```
does that exist? Google doesn't find it and neither does docs.pytorch.org --https://pytorch.org/docs/stable/search.html?q=is_cpu&check_keywords=yes&area=default

albanD replied:
Ho we have is_cuda but not is_cpu. interesting...

swolchok commented on torch/nn/modules/activation.py:
```diff
@@ -1034,6 +1056,66 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
+                    self.in_proj_bias,
+                    self.out_proj.weight,
+                    self.out_proj.bias,
+                    key_padding_mask if key_padding_mask is not None else attn_mask,
```
I believe I wrote that originally, but it doesn't work properly in TorchScript

swolchok commented on torch/nn/modules/transformer.py:
```diff
@@ -336,6 +364,53 @@ def forward(self, src: Tensor, src_mask: Optional[Tensor] = None, src_key_paddin
+                    self.self_attn.out_proj.weight,
+                    self.self_attn.out_proj.bias,
+                    self.activation_relu_or_gelu == 2,
+                    False, # norm_first, currently not supported
```
I could, but it would entail (a small amount of) extra overhead to no benefit since we know it's going to be False. If/when we add support, I helpfully mentioned `norm_first` so that we will see this potential use_site when searching for places to update in our text editor.

swolchok commented:
"MultiheadAttention does not support NestedTensor outside of its fast path. The fast path was not hit because some Tensor argument has_torch_function" on linux-bionic-py3.7-clang . Why the heck are there platform-specific torch_functions? I guess I will patch the test to check itself if there are torch_functions set up for some reason and do the right thing.

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

ezyang commented:
That must be the crossref config; it runs all the tests with a no-op torch function to check if our implementations invariant to a torch function mode. In your case, you should skip the fast path tests here.

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok committed:
lintrunner started working, appease it on "[PyTorch] Add native fast path for transformer encoder inference"


The current PyTorch multi-head attention and transformer
implementations are slow. This should speed them up for inference.

Differential Revision: [D35239925](https://our.internmc.facebook.com/intern/diff/D35239925/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35239925/)!

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge this please

cpuhrsch commented:
@pytorchbot merge this

swolchok commented:
@pytorchbot revert this

bigfootjon commented:
Manually reverted with my admin powers: https://github.com/pytorch/pytorch/commit/2387efd35601205eabeb0dfe90731aa1dbf0eabb
